File: Practicals/Practical 4/Game.java
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

// Helper class to simulate animations (optional)
class AnimationHelper {
    public static void animate(String action, int duration) {
        System.out.print(action);
        try {
            for (int i = 0; i < duration; ++i) {
                System.out.print(".");
                TimeUnit.MILLISECONDS.sleep(500);
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        System.out.println(" Done!");
    }
}

public class Game {
    private Section mainFarmSection;
    private List<Truck> trucks;

    public Game() {
        mainFarmSection = new Section("Main Farm");

        Section northSection = new Section("North Farm");
        Section southSection = new Section("South Farm");

        mainFarmSection.addFarmUnit(northSection);
        mainFarmSection.addFarmUnit(southSection);

        trucks = new ArrayList<>();
    }

    public void setup() {
        Section northSection = (Section) mainFarmSection.getUnits().get(0);
        Section southSection = (Section) mainFarmSection.getUnits().get(1);

        // Adding crop fields and applying Decorators
        BasicCropField cropField1 = new BasicCropField("Wheat", 100, new DrySoil());
        BasicCropField cropField2 = new BasicCropField("Corn", 120, new FruitfulSoil());
        BasicCropField cropField3 = new BasicCropField("Barley", 150, new FloodedSoil());

        Fertilizer fertilizerField1 = new Fertilizer(cropField1);
        ExtraBarn extraBarnField2 = new ExtraBarn(cropField2, 50);

        northSection.addFarmUnit(fertilizerField1);
        northSection.addFarmUnit(extraBarnField2);
        northSection.addFarmUnit(cropField3);

        BasicCropField cropField4 = new BasicCropField("Soy", 90, new DrySoil());
        southSection.addFarmUnit(cropField4);

        // Adding trucks
        DeliveryTruck deliveryTruck = new DeliveryTruck();
        FertilizerTruck fertilizerTruck = new FertilizerTruck();

        trucks.add(deliveryTruck);
        trucks.add(fertilizerTruck);

        fertilizerField1.addTruck(fertilizerTruck);
        extraBarnField2.addTruck(deliveryTruck);
    }

    public void start() {
        System.out.println("Game Started!");
        // AnimationHelper.animate("Loading farm details", 5);

        System.out.println("////////////Breadth First Traversal (BFT)//////////////////");
        BFT bft = new BFT(mainFarmSection);
        bft.firstFarm();
        while (!bft.isDone()) {
            FarmUnit currentFarm = bft.currentFarm();
            // AnimationHelper.animate("Checking farm " + currentFarm.getCropType(), 3);
            System.out.println("Farm: " + currentFarm.getCropType() + " (Capacity: " + currentFarm.getTotalCapacity() + ")");
            bft.next();
        }

        System.out.println("////////////Depth First Traversal (DFT)//////////////////");
        DFT dft = new DFT(mainFarmSection);
        dft.firstFarm();
        while (!dft.isDone()) {
            FarmUnit currentFarm = dft.currentFarm();
            // AnimationHelper.animate("Inspecting farm " + currentFarm.getCropType(), 2);
            System.out.println("Farm: " + currentFarm.getCropType() + " (Capacity: " + currentFarm.getTotalCapacity() + ")");
            dft.next();
        }

        // Simulate logistics operations with trucks
        System.out.println("////////////Logistics Operations//////////////////");
        // AnimationHelper.animate("Trucks are delivering fertilizer", 6);
        for (Truck truck : trucks) {
            truck.startEngine();
            truck.update();
        }

        // Applying Fertilizer
        System.out.println("Applying Fertilizer to all fields...");
        for (FarmUnit section : mainFarmSection.getUnits()) {
            if (section instanceof Section) {
                Section subSection = (Section) section;
                for (FarmUnit field : subSection.getUnits()) {
                    if (field instanceof BasicCropField) {
                        BasicCropField cropField = (BasicCropField) field;
                        cropField.checkSoil();
                        // AnimationHelper.animate("Fertilizing " + cropField.getCropType(), 4);
                    }
                }
            }
        }

        // Harvesting
        System.out.println("Harvesting crops...");
        for (FarmUnit section : mainFarmSection.getUnits()) {
            if (section instanceof Section) {
                Section subSection = (Section) section;
                for (FarmUnit field : subSection.getUnits()) {
                    if (field instanceof BasicCropField) {
                        BasicCropField cropField = (BasicCropField) field;
                        cropField.harvest(50); // Example yield
                        // AnimationHelper.animate("Harvesting " + cropField.getCropType(), 5);
                    }
                }
            }
        }

        // Display storage capacity after harvest
        System.out.println("Final storage capacities after harvest:");
        for (Truck truck : trucks) {
            System.out.println(truck.getTruckName());
        }
    }
}

File: Practicals/Practical 4/Game.cpp

#ifndef GAME_CPP
#define GAME_CPP

#include "Game.h"
#include <chrono>
#include <thread>
#include <iostream>


Game::Game() {
    mainFarmSection = std::make_shared<Section>("Main Farm");

    auto northSection = std::make_shared<Section>("North Farm");
    auto southSection = std::make_shared<Section>("South Farm");

    mainFarmSection->addFarmUnit(northSection);
    mainFarmSection->addFarmUnit(southSection);
}

void Game::setup() {
    auto northSection = std::dynamic_pointer_cast<Section>(mainFarmSection->getUnits()[0]);
    auto southSection = std::dynamic_pointer_cast<Section>(mainFarmSection->getUnits()[1]);

    // Adding crop fields and applying Decorators
    auto cropField1 = std::make_shared<BasicCropField>("Wheat", 100, std::make_shared<DrySoil>());
    auto cropField2 = std::make_shared<BasicCropField>("Corn", 120, std::make_shared<FruitfulSoil>());
    auto cropField3 = std::make_shared<BasicCropField>("Barley", 150, std::make_shared<FloodedSoil>());

    auto barn= std::make_shared<Barn>(20);
    


    auto fertilizerField1 = std::make_shared<Fertilizer>(cropField1);
    auto extraBarnField2 = std::make_shared<ExtraBarn>(cropField2, 50);

    fertilizerField1->increaseProduction();
    fertilizerField1-> harvestCrops(10);
    fertilizerField1->getCapacity();



    extraBarnField2->increaseProduction();
    extraBarnField2-> harvestCrops(10);
    
    extraBarnField2->getCapacity();
    

    northSection->addFarmUnit(fertilizerField1);
    northSection->addFarmUnit(extraBarnField2);
    northSection->addFarmUnit(cropField3);

    auto cropField4 = std::make_shared<BasicCropField>("Soy", 90, std::make_shared<DrySoil>());
    southSection->addFarmUnit(cropField4);
    southSection->addFarmUnit(cropField2);
    southSection->addFarmUnit(barn);

    // Adding trucks
    auto deliveryTruck = std::make_shared<DeliveryTruck>();
    auto fertilizerTruck = std::make_shared<FertilizerTruck>();

    trucks.push_back(deliveryTruck);
    trucks.push_back(fertilizerTruck);

    fertilizerField1->buyTruck(fertilizerTruck);
    fertilizerField1->buyTruck(deliveryTruck);
    extraBarnField2->buyTruck(deliveryTruck);
    // extraBarnField2->addTruck(fertilizerTruck);
    extraBarnField2->buyTruck(fertilizerTruck);

    
}

void Game::start() {
     std::cout<<"\n\n\n";
    std::cout << "Game Started!" << std::endl;
    // animate("Loading farm details");
    std::cout << "////////////Breadth First Traversal (BFT)//////////////////" << std::endl;

    BFT bft(mainFarmSection);
    // bft.firstFarm();
    while (!bft.isDone()) {
        auto currentFarm = bft.currentFarm();
        // animate("Checking farm " + currentFarm->getCropType(), 3);
        std::cout<< currentFarm->getCropType() << " (Capacity: " << currentFarm->getTotalCapacity() << ")" << std::endl;
        bft.next();
    }
    std::cout<<"\n\n\n";
    std::cout << "////////////Depth First Traversal (DFT)//////////////////" << std::endl;
    DFT dft(mainFarmSection);
    dft.firstFarm();
    while (!dft.isDone()) {
        auto currentFarm = dft.currentFarm();
        // animate("Inspecting farm " + currentFarm->getCropType(), 2);
        std::cout<< currentFarm->getCropType() << " (Capacity: " << currentFarm->getTotalCapacity() << ")" << std::endl;
        dft.next();
    }
    std::cout<<"\n\n\n";
    // Simulate logistics operations with trucks
    std::cout << "////////////Logistics Operations//////////////////" << std::endl;
    // animate("Trucks are delivering fertilizer", 6);
    for (auto& truck : trucks) {
        truck->startEngine();
        truck->update();
    }

    // Applying Fertilizer
    std::cout << "Applying Fertilizer to all fields..." << std::endl;
    for (auto& section : mainFarmSection->getUnits()) {
        auto subSection = std::dynamic_pointer_cast<Section>(section);
        if (subSection) {
            for (auto& field : subSection->getUnits()) {
                auto cropField = std::dynamic_pointer_cast<BasicCropField>(field);
                if (cropField) {
                    cropField->checkSoil();
                    // animate("Fertilizing " + cropField->getCropType(), 4);
                }
            }
        }
    }

     std::cout<<"\n\n\n";
    std::cout << "RAINY DAY" << std::endl;
    for (auto& section : mainFarmSection->getUnits()) {
        auto subSection = std::dynamic_pointer_cast<Section>(section);
        if (subSection) {
            for (auto& field : subSection->getUnits()) {
                auto cropField = std::dynamic_pointer_cast<BasicCropField>(field);
                if (cropField) {
                    std::cout<<"Field: "<<cropField->getCropType()<<std::endl;
                    cropField->getState()->rain(cropField);
                }
            }
        }
    }
    // Harvesting
     std::cout<<"\n\n\n";
    std::cout << "Harvesting crops..." << std::endl;
    for (auto& section : mainFarmSection->getUnits()) {
        auto subSection = std::dynamic_pointer_cast<Section>(section);
        if (subSection) {
            for (auto& field : subSection->getUnits()) {
                auto cropField = std::dynamic_pointer_cast<BasicCropField>(field);
                if (cropField) {
                    std::cout<<"Field: "<<cropField->getCropType()<<std::endl;
                    cropField->harvest(50);  
                    // animate("Harvesting " + cropField->getCropType(), 5);
                }
            }
        }
    }


    // Display storage capacity after harvest
    std::cout << "Trucks in storage: " << std::endl;
    for (auto& truck : trucks) {
        std::cout << truck->getTruckName()<< std::endl;

    }
    

    

}
#endif

